### PR TITLE
libretro: fix build with GCC 14

### PR DESCRIPTION
**Pull request by thiagokokada**
Fix `retroarch-full`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lucasewd** [comment]
# [nixpkgs-reviewd](https://github.com/nixosbrasil/nixpkgs-reviewd) workflow started

**Status and logs**: https://github.com/nixosbrasil/nixpkgs-reviewd/actions/runs/12530792928
**Extra args**: ``

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368955`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>libretro.fbalpha2012</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2016</li>
    <li>libretro.np2kai</li>
    <li>retroarch-free</li>
    <li>retroarch-full</li>
  </ul>
</details>

**Comment by lucasewd** [comment]
# [nixpkgs-reviewd](https://github.com/nixosbrasil/nixpkgs-reviewd) workflow finished

**Status and logs**: https://github.com/nixosbrasil/nixpkgs-reviewd/actions/runs/12530792928


## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368955`

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>libretro.fbalpha2012</li>
    <li>libretro.mame2003</li>
    <li>libretro.np2kai</li>
  </ul>
</details>

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368955`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>libretro.fbalpha2012</li>
    <li>libretro.mame2003</li>
    <li>libretro.np2kai</li>
  </ul>
</details>

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368955`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>libretro.fbalpha2012</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2016</li>
    <li>libretro.np2kai</li>
    <li>retroarch-free</li>
    <li>retroarch-full</li>
  </ul>
</details>

**Review by paparodeo — COMMENTED** [review]
it looks like, at least for the 2 I tried, that rather than pinning to gcc13 can add `env.NIX_CFLAGS_COMPILE = "-Wno-error=incompatible-pointer-types";` instead. still not great but slightly better than pinning to an old compiler / toolchain.

File: pkgs/applications/emulators/libretro/cores/fbalpha2012.nix [hunk 1]
```diff
@@ -1,11 +1,14 @@
 {
   lib,
+  gcc13Stdenv,
   fetchFromGitHub,
   mkLibretroCore,
 }:
 mkLibretroCore {
   core = "fbalpha2012";
   version = "0-unstable-2024-10-21";
+  # https://github.com/libretro/fbalpha2012/issues/116
```
Comment (paparodeo) [inline pkgs/applications/emulators/libretro/cores/fbalpha2012.nix:10]: this failure is a real bug. it assigns an `int*` to a `long*` (which is a crc32 table) then does uses the table to do encrypt / decryption using the wrong crc32 values.  the code vendored from zlib so it is easy enough to replace.

<details><summary>unvendor zlib and minizip</summary>

```diff
diff --git a/pkgs/applications/emulators/libretro/cores/fbalpha2012.nix b/pkgs/applications/emulators/libretro/cores/fbalpha2012.nix
index febf6f7935a8..aadf09cb7822 100644
--- a/pkgs/applications/emulators/libretro/cores/fbalpha2012.nix
+++ b/pkgs/applications/emulators/libretro/cores/fbalpha2012.nix
@@ -2,8 +2,10 @@
   lib,
   fetchFromGitHub,
   mkLibretroCore,
+  zlib,
+  runCommand
 }:
-mkLibretroCore {
+mkLibretroCore rec {
   core = "fbalpha2012";
   version = "0-unstable-2024-10-21";
 
@@ -14,8 +16,27 @@ mkLibretroCore {
     hash = "sha256-giEV09dT/e82bmDlRkxpkW04JcsEZc/enIPecqYtg3c=";
   };
 
+  sourceRoot = "${src.name}/svn-current/trunk";
+
+  # unvendor zlib and broken minizip code
+  postPatch = let
+    minizip-src = runCommand "minizip-src" {} ''
+      mkdir $out
+      unpackFile ${zlib.src}
+      cp */contrib/minizip/{unzip.*,ioapi.*,crypt.h} $out/
+    '';
+  in ''
+    substituteInPlace ${makefile} \
+      --replace-fail '-I$(FBA_LIB_DIR)/zlib' ""
+
+    cp ${minizip-src}/* src/burner
+  '';
+
+  buildInputs = [ zlib ];
+
+  makeFlags = [ "EXTERNAL_ZLIB=1" ];
+
   makefile = "makefile.libretro";
-  preBuild = "cd svn-current/trunk";
 
   meta = {
     description = "Port of Final Burn Alpha ~2012 to libretro";

```
</details>

**Comment by thiagokokada** [comment]
> it looks like, at least for the 2 I tried, that rather than pinning to gcc13 can add `env.NIX_CFLAGS_COMPILE = "-Wno-error=incompatible-pointer-types";` instead. still not great but slightly better than pinning to an old compiler / toolchain.

Done, now `np2kai` and `mame2003` are building with GCC 14.

The only one that is kept in GCC 13 is `mame2016`, that build failures are not related to compilation but link failures.

File: pkgs/applications/emulators/libretro/cores/mame2016.nix [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   lib,
+  gcc13Stdenv,
```
Comment (paparodeo) [inline pkgs/applications/emulators/libretro/cores/mame2016.nix:3]: for this one I get this error:
```
../../../../../3rdparty/rapidjson/include/rapidjson/document.h:317:82: error: assignment of read-only member 'rapidjson::GenericStringRef<CharType>::length'
  317 |     GenericStringRef& operator=(const GenericStringRef& rhs) { s = rhs.s; length = rhs.length; }
      |                                  
```
which doesn't break the build right away so have to search for it in the log.

rapidjson is a header only library so it is pretty easy to replace.

```diff
diff --git a/pkgs/applications/emulators/libretro/cores/mame2016.nix b/pkgs/applications/emulators/libretro/cores/mame2016.nix
index 31e929b7089a..5b3efdb1ea88 100644
--- a/pkgs/applications/emulators/libretro/cores/mame2016.nix
+++ b/pkgs/applications/emulators/libretro/cores/mame2016.nix
@@ -3,6 +3,7 @@
   alsa-lib,
   fetchFromGitHub,
   mkLibretroCore,
+  rapidjson,
   python3,
 }:
 mkLibretroCore {
@@ -16,6 +17,11 @@ mkLibretroCore {
     hash = "sha256-IsM7f/zlzvomVOYlinJVqZllUhDfy4NNTeTPtNmdVak=";
   };
 
+  postPatch = ''
+    rm -r 3rdparty/rapidjson
+    ln -s ${lib.getInclude rapidjson} 3rdparty/rapidjson
+  '';
+
   patches = [ ./patches/mame2016-python311.patch ];
   extraNativeBuildInputs = [ python3 ];
   extraBuildInputs = [ alsa-lib ];

```

built for me using gcc-14.  I noticed that there are a ton of things that could also get unvendored.
eg:
in the makefile I saw
```
# USE_SYSTEM_LIB_EXPAT = 1
  # USE_SYSTEM_LIB_ZLIB = 1
  # USE_SYSTEM_LIB_JPEG = 1
  # USE_SYSTEM_LIB_FLAC = 1
  # USE_SYSTEM_LIB_LUA = 1
  # USE_SYSTEM_LIB_SQLITE3 = 1
  # USE_SYSTEM_LIB_PORTMIDI = 1
  # USE_SYSTEM_LIB_PORTAUDIO = 1
  # USE_SYSTEM_LIB_UV = 1
  # USE_BUNDLED_LIB_SDL2 = 1
```
so could use nixpkgs for all of those too -- but that is out of scope for this PR.

**Review by paparodeo — APPROVED** [review]
LGTM

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368955`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>libretro.fbalpha2012</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2016</li>
    <li>libretro.np2kai</li>
    <li>retroarch-free</li>
    <li>retroarch-full</li>
  </ul>
</details>
